File: web/src/api/grammar.ts
import { get, post, ApiResponse } from '../utils/request'

export type GrammarOption = { key: string; text: string }
export type GrammarExample = { en: string; zh: string }

export type GrammarLessonListItem = {
  id: number
  title: string
  topic?: string
  level: string
  summary?: string
  estimatedMinutes?: number
  questionCount?: number
  lastScore?: number
  lastCorrectCount?: number
  lastQuestionCount?: number
  lastCompletedAt?: string
}

export type GrammarQuestionView = {
  id: number
  stem: string
  options: GrammarOption[]
  sortOrder?: number
}

export type GrammarLessonDetail = {
  id: number
  title: string
  topic?: string
  level: string
  explanation: string
  examples: GrammarExample[]
  summary?: string
  estimatedMinutes?: number
  questions: GrammarQuestionView[]
}

export type GrammarAnswerDetail = {
  questionId: number
  answer: string
  correct: boolean
  rightAnswer?: string
  stem?: string
  explanation?: string
}

export type GrammarSubmitResult = {
  recordId: number
  lessonId: number
  title: string
  topic?: string
  level: string
  questionCount: number
  correctCount: number
  score: number
  durationSec: number
  completedAt?: string
  details: GrammarAnswerDetail[]
}

export const listGrammarLessons = (params?: {
  level?: string
  page?: number
  pageSize?: number
}): Promise<ApiResponse<{ list: GrammarLessonListItem[]; total: number }>> => {
  return get('/grammar/lessons', { params })
}

export const getGrammarLesson = (id: number): Promise<ApiResponse<GrammarLessonDetail>> => {
  return get(`/grammar/lessons/${id}`)
}

export const submitGrammarLesson = (
  id: number,
  data: { answers: Array<{ questionId: number; answer: string }>; durationSec?: number }
): Promise<ApiResponse<GrammarSubmitResult>> => {
  return post(`/grammar/lessons/${id}/submit`, data)
}

File: internal/models/grammar.go
package models

import (
	"time"

	common "github.com/LingByte/ling-base/common"
)

const (
	GrammarStatusDraft     = "draft"
	GrammarStatusPublished = "published"
)

// GrammarLesson 语法解析课
type GrammarLesson struct {
	common.BaseModel
	Title            string `json:"title" gorm:"size:256;not null;index;comment:课标题"`
	Topic            string `json:"topic" gorm:"size:128;index;comment:语法主题"`
	Level            string `json:"level" gorm:"size:32;index;comment:难度 初阶/中阶/高阶"`
	Explanation      string `json:"explanation" gorm:"type:text;not null;comment:语法讲解"`
	Examples         string `json:"examples" gorm:"type:text;comment:例句 JSON [{en,zh}]"`
	Summary          string `json:"summary" gorm:"size:512;comment:摘要"`
	Status           string `json:"status" gorm:"size:32;index;default:draft;comment:draft/published"`
	EstimatedMinutes int    `json:"estimatedMinutes" gorm:"default:5;comment:预计分钟"`
	SortOrder        int    `json:"sortOrder" gorm:"default:0;index;comment:排序"`
}

func (GrammarLesson) TableName() string { return "grammar_lessons" }

// GrammarQuestion 语法练习题
type GrammarQuestion struct {
	common.BaseModel
	LessonID    uint   `json:"lessonId" gorm:"index;not null;comment:课ID"`
	Stem        string `json:"stem" gorm:"type:text;not null;comment:题干"`
	Options     string `json:"options" gorm:"type:text;not null;comment:选项 JSON [{key,text}]"`
	Answer      string `json:"answer" gorm:"size:8;not null;comment:正确答案 key"`
	Explanation string `json:"explanation" gorm:"type:text;comment:解析"`
	SortOrder   int    `json:"sortOrder" gorm:"default:0;index;comment:题序"`
}

func (GrammarQuestion) TableName() string { return "grammar_questions" }

// GrammarRecord 语法练习答题记录
type GrammarRecord struct {
	common.BaseModel
	UserID        uint       `json:"userId" gorm:"index;not null;comment:用户ID"`
	LessonID      uint       `json:"lessonId" gorm:"index;not null;comment:课ID"`
	Answers       string     `json:"answers" gorm:"type:text;comment:答题快照 JSON"`
	QuestionCount int        `json:"questionCount" gorm:"comment:题目数"`
	CorrectCount  int        `json:"correctCount" gorm:"comment:答对数量"`
	Score         int        `json:"score" gorm:"comment:得分百分比 0-100"`
	DurationSec   int        `json:"durationSec" gorm:"default:0;comment:用时秒"`
	IsLatest      bool       `json:"isLatest" gorm:"default:false;index;comment:该课最新一次"`
	CompletedAt   *time.Time `json:"completedAt" gorm:"comment:完成时间"`
}

func (GrammarRecord) TableName() string { return "grammar_records" }
